feat(oracle): Add support for BULK COLLECT INTO

mkmoisen opened the pull request:
This commit adds support for the Oracle specific `BULK COLLECT INTO` syntax.

```
SELECT foo, bar
BULK COLLECT INTO l_rows
FROM bar
```

tobymao commented:
can you please link the relevant documentation from oracle?

tobymao commented on sqlglot/expressions.py:
```diff
@@ -2109,6 +2109,10 @@ class Into(Expression):
 
 
+class BulkCollectInto(Expression):
+    arg_types = {"this": True}
```
this is not necessary because it's identical to the base class

tobymao commented on sqlglot/tokens.py:
```diff
@@ -221,6 +221,7 @@ class TokenType(AutoName):
     AUTO_INCREMENT = auto()
     BEGIN = auto()
     BETWEEN = auto()
+    BULK_COLLECT_INTO = auto()
```
Let's not add a token, instead do match texts, we generally don't add tokens unless we have to.

tobymao commented on sqlglot/parser.py:
```diff
@@ -3195,6 +3199,12 @@ def _parse_into(self) -> t.Optional[exp.Into]:
         )
 
+    def _parse_bulk_collect_into(self) -> t.Optional[exp.BulkCollectInto]:
+        if not self._match(TokenType.BULK_COLLECT_INTO):
```
change this to match_text_seq

tobymao commented on sqlglot/parser.py:
```diff
@@ -3195,6 +3199,12 @@ def _parse_into(self) -> t.Optional[exp.Into]:
+        if not self._match(TokenType.BULK_COLLECT_INTO):
+            return None
+
+        return self.expression(exp.BulkCollectInto, this=self._parse_table(schema=True))
```
this doesnt' seem correct, it looks it can be a list of stuff right?

mkmoisen replied:
@tobymao I'm pretty sure it is always into a single variable, not a list. I will confirm tomorrow.

tobymao commented on sqlglot/parser.py:
```diff
@@ -2969,6 +2969,10 @@ def _parse_select(
 
+            bulk_collect_into = self._parse_bulk_collect_into()
+            if bulk_collect_into:
+                this.set("bulk_collect_into", bulk_collect_into)
```
you need to to make bulk_collect_into a first class property of select

tobymao replied:
can you have both into and bulk_into? if not, then i'd just use the into statement, perhaps you could just overide the into parser of oracle

mkmoisen replied:
@tobymao No, you can only have INTO or BULK COLLECT INTO, not both.

However, from an AST diff perspective, I think it would be useful to be able to make a distinction between these two, as they are semantically different.

mkmoisen commented:
@tobymao 

> can you please link the relevant documentation from oracle?

https://docs.oracle.com/en/database/oracle/oracle-database/19/lnpls/SELECT-INTO-statement.html

https://docs.oracle.com/en/database/oracle/oracle-database/19/lnpls/RETURNING-INTO-clause.html#GUID-38F735B9-1100-45AF-AE71-18FB74A899BE__CJAIAGHJ

Should this be added to a comment in the code anywhere?

tobymao commented on sqlglot/generator.py:
```diff
@@ -2464,6 +2470,10 @@ def select_sql(self, expression: exp.Select) -> str:
             into.pop()
 
+        bulk_collect_into = expression.args.get("bulk_collect_into")
+        if not self.SUPPORTS_SELECT_BULK_COLLECT_INTO and bulk_collect_into:
```
let's delete this section for now, it's not necessary

tobymao commented on sqlglot/generator.py:
```diff
@@ -2507,6 +2517,7 @@ def select_sql(self, expression: exp.Select) -> str:
             expression,
             f"SELECT{top_distinct}{kind}{expressions}",
             self.sql(expression, "into", comment=False),
+            self.sql(expression, "bulk_collect_into", comment=False),
```
this may not be necessary if you add it as a special type of into

tobymao commented:
> @tobymao
> 
> > can you please link the relevant documentation from oracle?
> 
> https://docs.oracle.com/en/database/oracle/oracle-database/19/lnpls/SELECT-INTO-statement.html
> 
> https://docs.oracle.com/en/database/oracle/oracle-database/19/lnpls/RETURNING-INTO-clause.html#GUID-38F735B9-1100-45AF-AE71-18FB74A899BE__CJAIAGHJ
> 
> Should this be added to a comment in the code anywhere?

yea, you can perhaps where you parse bulk into

tobymao reviewed (changes requested):
thanks for the first PR. i've done an initial pass

georgesittas commented on sqlglot/expressions.py:
```diff
@@ -2106,7 +2106,7 @@ class PrimaryKey(Expression):
 # https://docs.aws.amazon.com/redshift/latest/dg/r_SELECT_INTO.html#r_SELECT_INTO-examples
 class Into(Expression):
-    arg_types = {"this": True, "temporary": False, "unlogged": False}
+    arg_types = {"this": True, "temporary": False, "unlogged": False, "bulk_collect_into": False}
```
[Nit] I'd rename this to just `bulk_collect` since "into" is encode din the expression's name too.

mkmoisen replied:
Done.

georgesittas reviewed (commented):
Toby's suggestion makes sense to me, given there are aliasing issues. Other than that, LGTM 👍

mkmoisen commented:
Hi @tobymao @georgesittas 

Please review my latest commit when you have a moment.

To make this BULK COLLECT INTO only a Oracle only Keyword, I did the following:

In dialects/oracle.py:Oracle:Tokenizer:Keywords I added: 

    "BULK COLLECT INTO": TokenType.BULK_COLLECT_INTO

However in `tokens.py:TokenType` I added:

    `BULK_COLLECT_INTO = auto()`
    
I'm not clear if you wanted this TokenType to be in Oracle specifically. I could not find any subclasses of TokenType in the project, so I assume this is ok. Please let me know if you want something else here.

georgesittas commented:
Actually, one final point, Toby [mentioned](https://github.com/tobymao/sqlglot/pull/4181#discussion_r1782121472) this as well and I'd like to double check it: Oracle's [docs](https://docs.oracle.com/en/database/oracle/oracle-database/19/lnpls/RETURNING-INTO-clause.html#GUID-38F735B9-1100-45AF-AE71-18FB74A899BE__CJAIAGHJ) show that you can have a csv after the `BULK COLLECT INTO` sequence. That means multiple values can be accepted. Did you test this?

mkmoisen commented:
> Actually, one final point, Toby [mentioned](https://github.com/tobymao/sqlglot/pull/4181#discussion_r1782121472) this as well and I'd like to double check it: Oracle's [docs](https://docs.oracle.com/en/database/oracle/oracle-database/19/lnpls/RETURNING-INTO-clause.html#GUID-38F735B9-1100-45AF-AE71-18FB74A899BE__CJAIAGHJ) show that you can have a csv after the `BULK COLLECT INTO` sequence. That means multiple values can be accepted. Did you test this?

@georgesittas I've been working with Oracle for 15 years and never knew BULK COLLECT could do this :).

However, this is also a problem with `INTO` for both Postgres and Oracle. Today sqlglot won't allow you to parse a csv and will only allow one value.

I think the core issue here is that `Into` is overloaded in Postgresql, with two cases, and sqlglot is currently only handling one case.

Case one, which sqlglot supports, is the Postgres CTAS alternative [here](https://www.postgresql.org/docs/current/sql-selectinto.html). This case is not available in Oracle.

Case two, which sqlglot does not support, is populating variables in PL/pgSQL (Postgresql) or in PL/SQL (Oracle)

----

My latest commit attempts to partially reconcile this. 

I have only enabled support for INTO with multiple values, as well as BULK COLLECT INTO with one or more values, in the Oracle dialect. 

Everything is working except for the case of transpiling Oracle into another language. It seems to me that attempts to do this should raise an error.

mkmoisen commented:
Initially, I attempted to have the `parser.py::Parser::_parse_into` handle both cases.

However, it does make the code harder to understand. I figured isolating INTO and BULK COLLECT INTO into the Oracle dialect would simplify the understanding.

In addition, I couldn't figure out how to both parse for the table (self._parse_table) as well as parse the CSV (self._parse_expressions). It kept having collisions.

georgesittas commented on sqlglot/dialects/oracle.py:
```diff
@@ -344,3 +356,7 @@ def queryoption_sql(self, expression: exp.QueryOption) -> str:
             return rename_func(func_name)(self, expression)
+
+        def into_sql(self, expression: exp.Into) -> str:
+            into = "INTO" if not expression.args.get("bulk_collect") else "BULK COLLECT INTO"
```
What about `TEMPORARY` and `UNLOGGED`? It seems like we're generating them in the base method. Are these needed for Oracle?

mkmoisen replied:
No, `TEMPORARY` and `UNLOGGED` are not used in Oracle's INTO statement.

These are only used in the CTAS case in postgresql.

mkmoisen replied:
> Perhaps a workaround is to preserve the previous implementation, i.e. set this instead of expressions, if there is only one expression.

@georgesittas Would you please give a suggestion on how to accomplish this?

I tried but was unable to figure it out.

```
expressions = self._parse_expressions()
if len(expressions) == 1:
    ...
```

I think i would need to somehow rewind and then call:

    self._parse_table(schema=True)

georgesittas replied:
You can check if `bulk_collect_into` is truthy, in which case you'll set `expressions` to `_parse_csv(self._parse_bitwise)` (*), otherwise just set `this` to `self._parse_table(schema=True)`.

(*): `_parse_expressions` consumes more things and is probably overkill here.

mkmoisen replied:
Hi @georgesittas ,

For Oracle, INTO and BULK COLLECT INTO can both support multiple values in a csv.

This latest commit seems to solve the issue. If only one expression is found, then I call self._advance(-1), and then revert to the previous implementation.

georgesittas reviewed (commented):
> Case two, which sqlglot does not support, is populating variables in PL/pgSQL (Postgresql) or in PL/SQL (Oracle)

That's fine, procedural SQL code (e.g. variable population as you mention) isn't prioritized in SQLGlot anyway.

> Everything is working except for the case of transpiling Oracle into another language. It seems to me that attempts to do this should raise an error.

Perhaps a workaround is to preserve the previous implementation, i.e. set `this` instead of `expressions`, if there is only one expression.

mkmoisen commented:
Hi @georgesittas 

Please see my recent commit. I think it is mostly there. If only one value is provided for INTO or BULK COLLECT INTO, it will revent to the previous implemenation. This way, the statement can be transpiled into mysql CTAS, or postgres/tsql SELECT INTO, so long as it only has one value.

However, the case where the user has more than one value is a bit more complicated.

It wouldn't make sense to transpile this oracle statement into mysql CTAS:

    select foo, bar into l_foo, l_bar from baz

I feel like it should raise an error.

----

When it comes to postgresql, it's a bit more sketchy because of the two cases I mentioned above.

This is a totally valid Oracle statement that could be transpiled into postgresql for plpgsql:

    select foo, bar into l_foo, l_bar from baz
    
Yet this is totally invalid to transpile it to postgreql's CTAS alternative SELECT INTO.

mkmoisen commented:
I see two approaches:

## Assume Postgresql's INTO is the CTAS alternative case, not the Pl/pgSQL case

Here, an oracle INTO with one value can be converted into postgresq/mysql/tsl. (Although I feel like this is semantically incorrect, since Oracle's INTO is never a CTAS).

But an oracle INTO with more than one value should raise an exception if attempting to transpile it to postgres, mysql, tsql, etc.

## Add support for multiple values in Postgresql INTO to handle the pl/pgSQL case

If a postgresql INTO uses more than one variable, it should raise an exception if attempting to transpile to mysql, tsql, etc. However it could be transpiled to Oracle.

Oracle INTO's that use more than one variable can be transpiled into postgresql but nothing else. 



As an aside, it seems to me that any Oracle INTO statement should not be allowed to be transpiled to any dialect other than postgresql. Semantically it doesn't make sense, as Oracle INTO is not a CTAS alterantive. It's only a CTAS alternative in postgres/tsql.

mkmoisen commented:
As an aside, it seems to me that any Oracle INTO statement should not be allowed to be transpiled to any dialect other than postgresql. Semantically it doesn't make sense, as Oracle INTO is not a CTAS alternative. 

 Only in postgres/tsql is INTO an alternative for CTAS.

georgesittas reviewed (commented):
I think the `advance(-1)` approach isn't quite right, because you may have multiple tokens so you won't really backtrack completely.

I'll just get this PR to the finish line though, thanks!

georgesittas commented:
Re: transpilation - I wouldn't worry too much about it now, unless you have a use case of transpiling to postgres / others mentioned. We can improve it incrementally in followup PRs.

georgesittas commented:
See https://github.com/tobymao/sqlglot/commit/5a123a54ecd033c0a104e33476b17d816a09caac

mkmoisen commented:
Hi @georgesittas 

Thank you for the fix.

But I think there is still one open issue:
```
>>> import sqlglot

>>> parsed = sqlglot.parse_one('select foo, bar into l_foo, l_bar from baz', dialect='oracle')

>>> print(parsed.sql(dialect='postgres')
SELECT foo, bar INTO  FROM baz

>>> print(parsed.sql('mysql'))
CREATE TABLE  AS SELECT foo, bar FROM baz
```

Essentially, when transpiling an oracle INTO with more than one value to any other dialect, the table name won't be set.

This produces invalid SQL.

It seems to me that this should instead raise some kind of unsupported exception - definitely for mysql/tsql etc. For Postgres you could either do the same, or add support for multiple values in the INTO for postgres.

Thanks!
